chia/wallet/wallet_singleton_store.py

```diff
@@ -0,0 +1,281 @@
+            await conn.execute(
+                f"INSERT or REPLACE INTO singletons ({columns}) VALUES({value_placeholders})",
+                (
+                    record.coin_id(),
+                    record.coin.parent_coin_info,
+                    record.coin.puzzle_hash,
+                    record.coin.amount,
+                    record.singleton_id,
+                    record.wallet_id,
+                    bytes(record.inner_puzzle),
+                    record.inner_puzzle_hash,
+                    int(record.confirmed),
+                    record.confirmed_at_height,
+                    record.spent_height,
+                    bytes(record.lineage_proof),
+                    record.generation,
+                    record.timestamp,
+                ),
+            )
+
+            if record.custom_data:
+                custom_data_columns = "coin_id, field_name, value"
+                coin_id = record.coin_id()
+                for key, val in record.custom_data.items():
+                    await conn.execute(
+                        f"INSERT or REPLACE INTO singleton_data ({custom_data_columns}) VALUES(?, ?, ?)",
+                        (coin_id, key, val),
+                    )
+
+    async def get_custom_data_by_coin_id(self, coin_id: bytes) -> Optional[Dict[str, Any]]:
+        async with self.db_wrapper.reader_no_transaction() as conn:
+            rows = await conn.execute_fetchall(
+                "SELECT field_name, value FROM singleton_data WHERE coin_id = ?", (coin_id,)
+            )
+        custom_data = {}
+        for row in rows:
+            custom_data[row[0]] = row[1]
+        if custom_data:
+            return custom_data
+        else:
+            return None
+
+    async def get_records_by_singleton_id(
+        self,
+        singleton_id: bytes32,
+        min_generation: Optional[uint32] = None,
+        max_generation: Optional[uint32] = None,
+        num_results: Optional[uint32] = None,
+    ) -> List[SingletonCoinRecord]:
+        """
+        Retrieves all stored singletons for a singleton ID.
+        """
+        query_params: List[Union[bytes32, uint32]] = [singleton_id]
+        for optional_param in (min_generation, max_generation, num_results):
+            if optional_param is not None:
+                query_params.append(optional_param)
+
+        async with self.db_wrapper.reader_no_transaction() as conn:
+            rows = await conn.execute_fetchall(
+                "SELECT * from singletons WHERE singleton_id=? "
+                f"{'AND generation >=? ' if min_generation is not None else ''}"
+                f"{'AND generation <=? ' if max_generation is not None else ''}"
+                "ORDER BY generation DESC"
+                f"{' LIMIT ?' if num_results is not None else ''}",
+                tuple(query_params),
+            )
+        records = []
+        for row in rows:
+            custom_data = await self.get_custom_data_by_coin_id(row[0])
```

Comment:
why not join these tables vs executing a separate SQL for each row?

Reply:
I couldn't find a way to use JOIN that would return a serialized custom_data, so if we did use a join we'd have to go through each row and serialize it anyway. Not sure whether this is considerably less efficient?

Reply:
the round-trip to the sqlite thread is expensive, so the fewer requests you make, the faster (typically).

Reply:
I think it would look something like this:
```
SELECT field_name, value FROM singleton_data WHERE coin_id in (
   SELECT coin_id from singletons
   WHERE singleton_id=?
   ORDER BY generation DESC
   LIMIT ?
)
```

Reply:
Thanks, I'll take this.